File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.java
class Solution { // tc is O(n2) and sc is O(1)
    public int numberOfSubstrings(String s) {
        int count = 0,left =0, n = s.length();
        int[] charCount = new int[3];
    
        for(int right = 0;right<n;right++){
            
         charCount[s.charAt(right)-'a']++;
            
            while(charCount[0] > 0 && charCount[1] > 0 && charCount[2] >0){
                count += n-right;
                charCount[s.charAt(left) - 'a']--;
                left++;
            }
            
               
           
        }
        return count;
    }
}

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int count = 0,left =0, n = s.size();
        int charCount[3] ={0};
    
        for(int right = 0;right<n;right++){
            
         charCount[s[right]-'a']++;
            
            while(charCount[0] > 0 && charCount[1] > 0 && charCount[2] >0){
                count += n-right;
                // shrking the window size
                charCount[s[left] - 'a']--;
                left++;
            }    
        }
        return count;
    }
};